File: strings/checkIfAllAsAppearsBeforeAllBs.java
// https://leetcode.com/problems/check-if-all-as-appears-before-all-bs/

package strings;

public class checkIfAllAsAppearsBeforeAllBs {
    public static void main(String[] args) {

        SolutionCIAAABAB solution = new SolutionCIAAABAB();

        System.out.println(solution.checkString("aab"));
        System.out.println(solution.checkString("abab"));
        System.out.println(solution.checkString("bbb"));

    }   
}


class SolutionCIAAABAB {

    // Submitted By @Jiganesh

    // TC: O(N)
    // SC: O(1)

    // Runtime: 1 ms, faster than 76.44% of Java online submissions for Check if All A's Appears Before All B's.
    // Memory Usage: 42.6 MB, less than 5.52% of Java online submissions for Check if All A's Appears Before All B's
    
    public boolean checkString(String s) {
        boolean occuredb = false;
        
        for (int i = 0; i<s.length(); i++){

            // check if any b has occured in string
            if (occuredb == false && s.charAt(i)=='b') occuredb = true;

            // check if b has occured and a has occured after it
            if (occuredb && s.charAt(i)=='a') return false;
        }
        return true;
    }
}

File: strings/checkIfAllAsAppearsBeforeAllBs.py
# Problem Link: https://leetcode.com/problems/check-if-all-as-appears-before-all-bs/submissions/

class Solution:
    def checkString(self, s: str) -> bool:
        last_seen = 'a'
        for i in s:
            if last_seen == 'a' and i == 'b':
                last_seen = 'b'
            elif last_seen == 'b' and i == 'a':
                return False
        return True
#         In this problem, we set the last seen element, initially as 'a'
#         If the next element we see is 'a', we move ahead,
#         else if the next element we see in 'b', we set the last seem to be 'b'
#         If my last seen in 'b' and I encounter 'a', I return False
       